test/plugin/base_test.rb

```diff
@@ -626,6 +630,87 @@ def test_struct_payload_json_container_log
     end
   end
 
+  # Verify that invalid tags are sanitized. Detailed implementation differs
+  # between the grpc and non-gprc paths, which is reflected in
+  # 'setup_logging_stubs_block' and 'verify_log_name_block'.
+  # When 'require_valid_tags' is true, we only accept string tags with valid
+  # characters.
+  def test_sanitize_invalid_tags_with_requires_valid_tags_true
+    setup_gce_metadata_stubs
+    {
+      'test' => 'test',
+      'germanß' => 'german%C3%9F',
+      'chinese中' => 'chinese%E4%B8%AD',
+      'specialCharacter/_-.' => 'specialCharacter%2F_-.'
+    }.each do |tag, converted_tag|
+      setup_logging_stubs(COMPUTE_PARAMS.merge(log_name: converted_tag)) do
+        @logs_sent = []
+        d = create_driver(REQUIRE_VALID_TAGS_CONFIG, tag)
+        d.emit('msg' => log_entry(0))
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS, 'structPayload') do
+        assert_equal "projects/#{PROJECT_ID}/logs/#{converted_tag}",
+                     @logs_sent[0]['logName']
+      end
+    end
+  end
+
+  # Verify that invalid tags are sanitized. Detailed implementation differs
+  # between the grpc and non-gprc paths, which is reflected in
+  # setup_logging_stubs_block and verify_log_name_block.
+  # When 'require_valid_tags' is false, we try to convert invalid tags by
+  # sanitizing invalid characters.
+  def test_sanitize_invalid_tags_with_requires_valid_tags_false
+    setup_gce_metadata_stubs
+    {
+      123 => '123',
+      'test' => 'test',
+      'germanß' => 'german%C3%9F',
+      'chinese中' => 'chinese%E4%B8%AD',
+      'specialCharacter/_-.#@*&^' => 'specialCharacter%2F_-.%23%40%2A%26%5E',
+      "nonutf8#{[0x92].pack('C*')}" => 'nonutf8%20',
+      [1, 2, 3] => '%5B1%2C%202%2C%203%5D',
+      { key: 'value' } => '%7B%22key%22%3D%3E%22value%22%7D',
+      "abc#{[0x92].pack('C*')}" => 'abc%20',
+      "#{[0x92].pack('C*')}" => '%20',
+      'abc@&^$*' => 'abc%40%26%5E%24%2A',
+      '@&^$*' => '%40%26%5E%24%2A'
+    }.each do |tag, converted_tag|
+      setup_logging_stubs(COMPUTE_PARAMS.merge(log_name: converted_tag)) do
+        @logs_sent = []
+        d = create_driver(APPLICATION_DEFAULT_CONFIG, tag)
+        d.emit('msg' => log_entry(0))
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS, 'structPayload') do
+        assert_equal "projects/#{PROJECT_ID}/logs/#{converted_tag}",
+                     @logs_sent[0]['logName']
```

Comment:
Reverted and added a comment to explain why we only verify log name in the first element.